content/courses/program-security/account-data-matching.md

```diff
@@ -76,214 +85,135 @@ pub struct AdminConfig {
-    }
-}
-
 #[derive(Accounts)]
 pub struct UpdateAdmin<'info> {
     #[account(
         mut,
-        has_one = admin
+        has_one = current_admin @ MyError::InvalidAdmin
     )]
     pub admin_config: Account<'info, AdminConfig>,
-    #[account(mut)]
-    pub admin: Signer<'info>,
-    pub new_admin: SystemAccount<'info>,
-}
-
-#[account]
-pub struct AdminConfig {
-    admin: Pubkey,
+    pub current_admin: Signer<'info>,
+    /// CHECK: This account is not read or written in this instruction
+    pub new_admin: UncheckedAccount<'info>,
 }
 ```
 
-Alternatively, you can use `constraint` to manually add an expression that must
-evaluate to true in order for execution to continue. This is useful when for
-some reason naming can’t be consistent or when you need a more complex
-expression to fully validate the incoming data.
+The `has_one = current_admin` constraint checks that the `current_admin`
+account's key matches the `admin` field in the `AdminConfig` account. If it
+doesn't match, Anchor will return the custom `InvalidAdmin` error.
+
+Alternatively, you can use the `constraint` attribute for more complex checks:
 
 ```rust
 #[derive(Accounts)]
 pub struct UpdateAdmin<'info> {
     #[account(
         mut,
-        constraint = admin_config.admin == admin.key()
+        constraint = admin_config.admin == current_admin.key() @ MyError::InvalidAdmin
     )]
     pub admin_config: Account<'info, AdminConfig>,
-    #[account(mut)]
-    pub admin: Signer<'info>,
-    pub new_admin: SystemAccount<'info>,
+    pub current_admin: Signer<'info>,
+    /// CHECK: This account is not read or written in this instruction
+    pub new_admin: UncheckedAccount<'info>,
 }
 ```
 
-## Lab
+This approach allows for more flexibility in your validation logic.
 
-For this lab we’ll create a simple “vault” program similar to the program we
-used in the Signer Authorization lesson and the Owner Check lesson. Similar to
-those labs, we’ll show in this lab how a missing data validation check could
-allow the vault to be drained.
+<Callout type="info">
+  Remember to define your custom errors in your program's error enum:
 
-#### 1. Starter
-
-To get started, download the starter code from the `starter` branch of
-[this repository](https://github.com/Unboxed-Software/solana-account-data-matching).
-The starter code includes a program with two instructions and the boilerplate
-setup for the test file.
-
-The `initialize_vault` instruction initializes a new `Vault` account and a new
-`TokenAccount`. The `Vault` account will store the address of a token account,
-the authority of the vault, and a withdraw destination token account.
+```rust
+#[error_code]
+pub enum MyError {
+    #[msg("Only the current admin can perform this action")]
+    InvalidAdmin,
+}
+```
 
-The authority of the new token account will be set as the `vault`, a PDA of the
-program. This allows the `vault` account to sign for the transfer of tokens from
-the token account.
+</Callout>
```

Comment:
Thanks @Epistetechnician ! I appreciate you actually read the CONTRIBUTING guide.

Reply:
This is the way ser!